File: src/2024/day11/2/main.go
package main

import (
	_ "embed"
	"strconv"
	"strings"
)

//go:embed data.txt
var fileContents string

func split(number uint64) []uint64 {
	if number == 0 {
		return []uint64{1}
	}

	numberStr := strconv.FormatUint(number, 10)
	if len(numberStr)%2 == 0 {
		half := len(numberStr) / 2
		return []uint64{
			mustParseUint(numberStr[:half]),
			mustParseUint(numberStr[half:]),
		}
	}
	return []uint64{number * 2024}
}

func mustParseUint(s string) uint64 {
	num, err := strconv.ParseUint(s, 10, 64)
	if err != nil {
		panic(err)
	}
	return num
}

func blinking(numbers []uint64, until int) int {
	stones := make(map[uint64]int)
	for _, n := range numbers {
		stones[n]++
	}

	for i := 0; i < until; i++ {
		newStones := make(map[uint64]int)

		for n, count := range stones {
			splits := split(n)
			for _, split := range splits {
				newStones[split] += count
			}
		}

		stones = newStones
	}
	total := 0
	for _, count := range stones {
		total += count
	}
	return total

}

func main() {
	line := strings.TrimSpace(strings.Split(fileContents, "\n")[0])
	numbers := make([]uint64, 0)
	for _, numStr := range strings.Fields(line) {
		num := mustParseUint(numStr)
		numbers = append(numbers, num)
	}

	total := blinking(numbers, 75)
	println(total)
}

File: src/2024/day11/2/main.rs
use std::collections::HashMap;
use std::fs;
use std::path::Path;

fn read_numbers() -> Vec<u64> {
    let current_file = Path::new(file!());
    let current_dir = current_file
        .parent()
        .expect("Failed to get parent directory");
    let file_path = current_dir.join("data.txt");

    fs::read_to_string(file_path)
        .expect("Failed to read file")
        .split_whitespace()
        .map(|num| num.parse().expect("Failed to parse number"))
        .collect()
}

fn split(number: u64) -> Vec<u64> {
    if number == 0 {
        return vec![1];
    }

    let number_str = number.to_string();
    if number_str.len() % 2 == 0 {
        let half = number_str.len() / 2;
        let first = number_str[..half].parse().unwrap();
        let second = number_str[half..].parse().unwrap();
        vec![first, second]
    } else {
        vec![number * 2024]
    }
}

fn blinking(numbers: Vec<u64>, until: usize) -> usize {
    let mut stones: HashMap<u64, usize> = HashMap::new();
    for &n in &numbers {
        *stones.entry(n).or_insert(0) += 1;
    }

    for _ in 0..until {
        let mut new_stones: HashMap<u64, usize> = HashMap::new();

        for (&n, &count) in stones.iter() {
            for split_num in split(n) {
                *new_stones.entry(split_num).or_insert(0) += count;
            }
        }

        stones = new_stones;
    }

    stones.values().sum()
}

fn main() {
    let numbers = read_numbers();
    let result = blinking(numbers, 75);

    println!("{}", result);
}
